### PR TITLE
fix: handle delay boundary add test for useInterval & useTimeout 

### PR DESCRIPTION
**Pull request by hchlq**
[[中文版模板 / Chinese template](https://github.com/alibaba/hooks/blob/master/.github/PULL_REQUEST_TEMPLATE/pr_cn.md)]

### 🤔 This is a ...

- [ ] New feature
- [x] Bug fix
- [ ] Site / documentation update
- [ ] Demo update
- [ ] TypeScript definition update
- [ ] Bundle size optimization
- [ ] Performance optimization
- [ ] Enhancement feature
- [ ] Internationalization
- [ ] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Other (about what?)

### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English |           |
| 🇨🇳 Chinese |           |

### ☑️ Self Check before Merge

⚠️ Please check all items below before review. ⚠️

- [ ] Doc is updated/provided or not needed
- [ ] Demo is updated/provided or not needed
- [ ] TypeScript definition is updated/provided or not needed
- [ ] Changelog is provided or not needed

### CODE REVIEW

File: packages/hooks/src/useInterval/__tests__/index.test.ts [hunk 1]
```diff
@@ -24,13 +24,26 @@ describe('useInterval', () => {
     expect(callback).toHaveBeenCalledTimes(3);
   });
 
-  it('delay is undefined should stop', () => {
-    let delay: number | undefined = undefined;
+  it('interval should stop', () => {
     const callback = jest.fn();
-    const { rerender } = setUp({ fn: callback, delay });
```
Comment (hchlq) [inline packages/hooks/src/useInterval/__tests__/index.test.ts:30]: type error 
image

File: packages/hooks/src/useTimeout/index.ts [hunk 2]
```diff
@@ -6,7 +6,11 @@ function useTimeout(fn: () => void, delay: number | undefined): void {
   const fnRef = useLatest(fn);
 
   useEffect(() => {
-    if (!isNumber(delay) || delay < 0) return;
+    if (!isNumber(delay) || delay < 0 || isNaN(delay)) {
```
Comment (crazylxr) [inline packages/hooks/src/useTimeout/index.ts:9]: 有了 isNumber 的判断，为什么还需要 isNaN 呢？我们并没有对 delay 做 number 的转换

Comment (hchlq) [reply packages/hooks/src/useTimeout/index.ts:9]: 我是这样想的，如果是NaN的话，应该也不支持启动定时器
typeof NaN 的结果是number, 所以对NaN处理一下

File: packages/hooks/src/useTimeout/index.ts [hunk 3]
```diff
@@ -2,11 +2,26 @@ import { useEffect } from 'react';
 import useLatest from '../useLatest';
 import { isNumber } from '../utils';
 
-function useTimeout(fn: () => void, delay: number | undefined): void {
+function useTimeout(
+  fn: () => void,
+  delay: number | undefined,
+  options?: {
+    immediate?: boolean;
```
Comment (brickspert) [inline packages/hooks/src/useTimeout/index.ts]: why `useTimeout` add `immediate` ?

Comment (hchlq) [reply packages/hooks/src/useTimeout/index.ts]: It mainly adapts to different scenes.
For example, I want to pop up toast at the beginning and end of the timer.
```typescript
  const [msg, updateMsg] = useState('开始');

  useTimeout(() => {
    alert(msg)
  }, 3000, {immediate: true});

  return <div onClick={() => updateMsg('结束')}>{msg}</div>;
```

Comment (hchlq) [reply packages/hooks/src/useTimeout/index.ts]: can you accept this immediate  parameter? If so, I'll update the docs

File: packages/hooks/src/useTimeout/index.ts (hunk 2 above)
Comment (hchlq) [reply packages/hooks/src/useTimeout/index.ts:9]: delay是NaN的情况

File: packages/hooks/src/useTimeout/index.ts (hunk 3 above)
Comment (brickspert) [reply packages/hooks/src/useTimeout/index.ts]: I don't think this attribute is needed.

Comment (hchlq) [reply packages/hooks/src/useTimeout/index.ts]: get it

**Review by brickspert — APPROVED** [review]
LGTM

**Comment by usercao** [comment]
image
image
image

@hchlq @crazylxr @brickspert 错误的校验现在疯狂警告⚠️

**Comment by crazylxr** [comment]
https://github.com/alibaba/hooks/releases/tag/v3.4.0  修复了，更新一下新的包

**Comment by usercao** [comment]
image

@crazylxr 建议恢复成`if (!isNumber(delay) || delay < 0) return;`

**Comment by hchlq** [comment]
都可以的哈，只是写法不一样而已

**Comment by usercao** [comment]
> 都可以的哈，只是写法不一样而已

作为一个明星开源项目肯定要追求代码的优雅啊，写法必须简洁清晰统一😓

**Comment by hchlq** [comment]
GET，这个会在 #1616 pr 修改哈

**Pull request merged**